sourcegen-generator-java/src/main/java/io/micronaut/sourcegen/JavaPoetSourceGenerator.java

```diff
@@ -154,8 +156,64 @@ private void writeEnum(Writer writer, EnumDef enumDef) throws IOException {
             enumBuilder.addAnnotation(asAnnotationSpec(annotation));
         }
 
-        for (String enumConstant : enumDef.getEnumConstants()) {
-            enumBuilder.addEnumConstant(enumConstant);
+        enumDef.getEnumConstants().forEach((name, exp) -> {
+            if (exp != null) {
+                enumBuilder.addEnumConstant(name, anonymousClassBuilder(renderExpression(null, null, exp)).build());
+            } else {
+                enumBuilder.addEnumConstant(name);
+            }
+        });
+
+        for (PropertyDef property : enumDef.getProperties()) {
+            TypeName propertyType = asType(property.getType(), enumDef);
+            String propertyName = property.getName();
+            FieldSpec.Builder fieldBuilder = FieldSpec.builder(
+                propertyType,
+                propertyName
+            ).addModifiers(Modifier.PRIVATE, Modifier.FINAL);
+            for (AnnotationDef annotation : property.getAnnotations()) {
+                fieldBuilder.addAnnotation(
+                    asAnnotationSpec(annotation)
+                );
+            }
+            property.getJavadoc().forEach(fieldBuilder::addJavadoc);
+            enumBuilder.addField(
+                fieldBuilder
+                    .build()
+            );
+            String capitalizedPropertyName = NameUtils.capitalize(propertyName);
+            enumBuilder.addMethod(MethodSpec.methodBuilder("get" + capitalizedPropertyName)
+                .addModifiers(property.getModifiersArray())
+                .addModifiers(Modifier.PUBLIC)
+                .returns(propertyType)
+                .addStatement("return this." + propertyName)
+                .build());
+        }
+
+        for (FieldDef field : enumDef.getFields()) {
+            FieldSpec.Builder fieldBuilder = FieldSpec
+                .builder(
+                    asType(field.getType(), enumDef),
+                    field.getName())
+                .addModifiers(field.getModifiersArray())
+                .addModifiers(Modifier.PRIVATE, Modifier.FINAL);
+            field.getInitializer().ifPresent(init ->
+                fieldBuilder.initializer(renderExpression(
+                    null,
+                    null,
+                    init
+                ))
+            );
+            field.getJavadoc().forEach(fieldBuilder::addJavadoc);
+            for (AnnotationDef annotation : field.getAnnotations()) {
+                fieldBuilder.addAnnotation(
+                    asAnnotationSpec(annotation)
+                );
+            }
+            enumBuilder.addField(
+                fieldBuilder
+                    .build()
+            );
```

Comment:
To avoid duplicaton of this with ClassDef, it would probably make sense to have separate `buildProperties` and `buildFields` methods that are called for both of them. `TypeSpec.Builder` seems to be the same.